File: notebookspython/03-urlify.py
class Solution:
    def replaceSpaces(self, chars: str, true_length: int) -> str:
        """Replace spaces in the true string section with %20."""
        s_list = list(chars)
        space_count = 0

        for i in range(true_length):
            if chars[i] == " ":
                space_count += 1

        new_index = true_length + space_count * 2

        for i in range(true_length - 1, -1, -1):
            if s_list[i] == " ":
                s_list[new_index - 1] = "0"
                s_list[new_index - 2] = "2"
                s_list[new_index - 3] = "%"
                new_index -= 3
            else:
                s_list[new_index - 1] = s_list[i]
                new_index -= 1

        return "".join(s_list).strip()


def urlify_pythonic(s: str, true_length: int) -> str:
    return s[:true_length].replace(" ", "%20")


if __name__ == "__main__":
    sol = Solution()

    assert sol.replaceSpaces("Mr John Smith    ", 13) == "Mr%20John%20Smith"
    assert sol.replaceSpaces("a b  ", 3) == "a%20b"
    assert sol.replaceSpaces("abc", 3) == "abc"
    assert urlify_pythonic("Mr John Smith    ", 13) == "Mr%20John%20Smith"

    print("All test cases passed ✅")

File: notebooksc#/03-urlify.cs
using System;
using System.Diagnostics;

public class Solution03URLify
{
    // Solution 1: In-place replacement with proper space allocation
    public string ReplaceSpaces(string chars, int trueLength)
    {
        char[] charArray = chars.ToCharArray();
        int spaceCount = 0;

        // Count spaces in the true string section
        for (int i = 0; i < trueLength; i++)
        {
            if (charArray[i] == ' ')
                spaceCount++;
        }

        int newIndex = trueLength + spaceCount * 2;

        // Replace spaces from right to left
        for (int i = trueLength - 1; i >= 0; i--)
        {
            if (charArray[i] == ' ')
            {
                charArray[newIndex - 1] = '0';
                charArray[newIndex - 2] = '2';
                charArray[newIndex - 3] = '%';
                newIndex -= 3;
            }
            else
            {
                charArray[newIndex - 1] = charArray[i];
                newIndex--;
            }
        }

        return new string(charArray, 0, trueLength + spaceCount * 2);
    }

    // Solution 2: Simple approach using built-in Replace
    public string URLifySimple(string s, int trueLength)
    {
        return s.Substring(0, trueLength).Replace(" ", "%20");
    }

    static void Main()
    {
        var sol = new Solution03URLify();

        Debug.Assert(sol.ReplaceSpaces("Mr John Smith    ", 13) == "Mr%20John%20Smith");
        Debug.Assert(sol.ReplaceSpaces("a b  ", 3) == "a%20b");
        Debug.Assert(sol.ReplaceSpaces("abc", 3) == "abc");
        Debug.Assert(sol.URLifySimple("Mr John Smith    ", 13) == "Mr%20John%20Smith");

        Console.WriteLine("All test cases passed ✅");
    }
}
